integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java

```diff
@@ -250,20 +251,30 @@ public void requestContainers() throws Exception {
     // is initialized to the number of requests made. (2) is then achieved by counting down whenever
     // a container is allocated, and waiting here for the number of outstanding requests to hit 0.
     int round = 0;
-    while (mWorkerHosts.size() < mNumWorkers && round < MAX_WORKER_CONTAINER_REQUEST_ROUNDS) {
-      requestWorkerContainers();
-      LOG.info("Waiting for {} worker containers to be allocated",
-          mOutstandingWorkerContainerRequestsLatch.getCount());
-      // TODO(andrew): Handle the case where something goes wrong and some worker containers never
-      // get allocated. See TACHYON-1410
-      mOutstandingWorkerContainerRequestsLatch.await();
-      round ++;
+    while (mWorkerHosts.size() < mNumWorkers && round ++ <
+        Constans.INTEGRATION_YARN_WORKER_CONTAINER_REQUEST_ROUNDS_MAX) {
+      int retries = 0;
+      boolean allocated = false;
+      while (retries ++ < Constants.INTEGRATION_YARN_CONTAINER_REQUEST_RETRY_COUNT) {
+        requestWorkerContainers();
+        LOG.info("Waiting for {} worker containers to be allocated",
+            mOutstandingWorkerContainerRequestsLatch.getCount());
+        allocated = mOutstandingWorkerContainerRequestsLatch.await((mNumWorkers － mWorkerHosts
+            .size()) * Constants.INTEGRATION_YARN_CONTAINER_REQUEST_TIMEOUT_S, TimeUnit.SECONDS)
```

Comment:
It's fine to just wait for the specified timeout - no need to multiply by `mNumWorkers - mWorkerHosts.size()`. The requests are all made in parallel, and the await gets confusing when `mNumWorkers - mWorkerHosts.size() == 0`.